### DISCUSSION TITLE
Disable previous steps when rendering navigation

### DISCUSSION THREAD

**Discussion by thomas-serverhouse** [spatie/laravel-livewire-wizard#32] [General]
Hi guys, 

I'm trying to create a breadcrumb using what I found here: https://spatie.be/docs/laravel-livewire-wizard/v1/usage/rendering-navigation

Here's how my form should work (made simple):
step1 : The user indicates if he is filling the form for himself or for a company 
step2 : The user gives details about his house (if he is filling the form for himself) 
step3 : A contact form 

When the user is filling for a "company" he goes from step1 to step3, but, how can i disable step2 in my breadcrumb?

I tried adding an attribute in the stepInfo method that i could use in my breadcrumb to check if the step can be accessed or not.
To do that i need the state of the previous steps, however when i check the previous states attributes with $this->state()->forStep('step1') it's all empty.

Is that a bug? Am i using it wrong? Do you guys have any workaround?

**Comment by ju5t** [comment]
This looks right to me. Do you have a test repo or a failing test perhaps?

I've PR'ed a passing test that shows how I would use it. See PR #33. The PR has some logic that's required for testing these wizards without javascript. Using `$this->state()->forStep()` should work; considering you're using the correct alias for that step. Have you checked that?

**Comment by thomas-serverhouse** [comment]
Thanks for the reply and the test.
I agree with you and the test provided, I can get the previous state with the state()->forStep() method when i'm inside the mount() method for example.

What I wanted to do though was getting the state within the stepInfo() method, when i'm inside it the previous states are not available, i'm guessing it's because the component is not mounted at this point. 
I'll try to make a test case later to better explain what i tried to do.

However I did manage to find a simple workaround (without using the wizard's navigation).
I created my own BreadcrumbTrait with a property i initialise inside the mount() method, it's not the best but i guess it'll do the job for now.

**Reply by ju5t** [reply]
We're mounting traits in the correct order as the `StepAware` trait, that uses `stepInfo()`, is booted after the component is mounted. Can you share the trait you've made so others can learn from it? Or perhaps it's something we can easily adjust in the wizard so it works out of the box?

To be sure, is this an example of you would like to use it?

```php
public function stepInfo(): array
{
    return [
        'label' => 'Your cart',
        'icon' => 'fa-shopping-cart',
        'state' => $this->state()->forStep('step-name'),
    ];
}
```

**Comment by thomas-serverhouse** [comment]
If i use the example i provided on my first post here's what i wanted to do:
```php
// Inside step2 component
public function stepInfo(): array
{
    $clickable = isset($this->state()->forStep('step1')['type']) && $this->state()->forStep('step1')['type'] !== 'company'
    return [
        'label' => 'My step 2',
        'icon' => 'fa-shopping-cart',
        'clickable' => $clickable,
    ];
}
```

And then in my breadcrumb
```php
@if ($step->isPrevious() && $step->clickable)
    wire:click="{{ $step->show() }}"
@endif
```

But this->state()->forStep('step1') always give me an empty array

**Reply by ju5t** [reply]
Hm, ok. I would've expected to have access to state as well within the `stepInfo` method, but clearly something is not working.

Could you perhaps try the following, just to check if the order of things and/or the trait is causing the issue:

1. Reimplement your previous `stepInfo` method, where you fetch the state from step one.
2. Inside the `mount` method of your second step, run `$this->bootedStepAware()`.

This would populate `$this->steps` twice; and it might just work. If it does, I'm guessing the order in which things run is the reason it returns an empty array.

I'm also curious what `$this->state()` returns. Can you share that too?

**Comment by thomas-serverhouse** [comment]
Here's the trait i wrote as a workaround

```php
trait BreadcrumbTrait
{

    public array $breadcrumbs;

    /**
     * Initialise the breadcrumb
     * Must be called in the component's mount() method
     * @return void
     */
    public function initBreadcrumb()
    {
        $this->breadcrumbs = [];

        $is_previous = true;
        foreach (MainComponent::STEPS as $step) {
            $is_clickable = $is_previous; // Previous step : it can be clicked
            $is_current   = false;

            // If it's the current step
            if (is_a($this, $step['class'])) {
                $is_clickable = false;
                $is_previous  = false;
                $is_current   = true;
            }

            // If it's a previous step : is there something more to check
            if ($is_previous) {
                if (/* something */) {
                    $is_clickable = /* */;
                }
            }

            $this->breadcrumbs[] = [
                'name'        => $step['name'],
                'label'        => $step['label'],
                'is_current'   => $is_current,
                'is_clickable' => $is_clickable,
            ];
        }

    }
}

```

Here's how my steps are defined
```php
// inside MainComponent
const STEPS = [
    [
        'name' => 'step1',
        'label' => 'My step 1',
        'class' => Step1Component::class
    ],
    // ....
]
```

Here's my breadcrumb
```php
@foreach($breadcrumbs as $step)
    <li @if ($step['is_clickable']) wire:click="showStep('{{$step['alias']}}')" @endif>
        {{ $step['label'] }}
    </li>
@endforeach
```
